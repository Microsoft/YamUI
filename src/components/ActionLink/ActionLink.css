.y-actionLink--wrapper {
  padding: 0.2rem 0 0.6rem;
}

.y-actionLink--icon {
  margin-top: 0.4rem;

  /* Remove when this is addressed: https://github.com/Microsoft/YamUI/issues/327 */
  position: static;
  top: 0;
}
