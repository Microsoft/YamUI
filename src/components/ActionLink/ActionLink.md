### Examples


With onClick (button implementation):

```js { "props": { "data-description": "with onClick", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"}]" } }
const Icon = require('../Icon/icons/Download').default;

<div>
  <ActionLink text="Download" icon={Icon} ariaLabel="Download this file" onClick={action('ActionLink clicked')} />
</div>
```

With long text and onClick:

```js { "props": { "data-description": "with onClick and long text" } }
const Icon = require('../Icon/icons/Upload').default;
const text = 'Upload a new version of this file';

<div style={{ maxWidth: '200px' }}>
  <ActionLink text={text} icon={Icon} ariaLabel={text} onClick={action('ActionLink clicked')} />
</div>
```

With href (link implementation):

```js { "props": { "data-description": "with href" } }
const Icon = require('../Icon/icons/Group').default;

<div>
  <ActionLink text="Group Info" icon={Icon} ariaLabel="Go to group info page" href="https://www.yammer.com" newWindow={true} />
</div>
```

With long text and href:

```js { "props": { "data-description": "with href and long text" } }
const Icon = require('../Icon/icons/SharePointLogoInverse').default;
const text = 'SharePoint Document Library website';

<div style={{ maxWidth: '200px' }}>
  <ActionLink text={text} icon={Icon} ariaLabel={text} href="https://www.yammer.com" newWindow={true} />
</div>
```

Compact:

```js { "props": { "data-description": "with compact" } }
const Like = require('../Icon/icons/Like').default;

<div>
  <ActionLink text="Like" icon={Like} ariaLabel="Like this Post" compact={true} />
</div>
```
