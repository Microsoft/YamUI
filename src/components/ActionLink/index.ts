/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default } from './ActionLink';
export * from './ActionLink.types';
