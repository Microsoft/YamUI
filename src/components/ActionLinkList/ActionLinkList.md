### Examples

Basic example:

```js { "props": { "data-description": "basic" } }
const Download = require('../Icon/icons/Download').default;
const Upload = require('../Icon/icons/Upload').default;
const Group = require('../Icon/icons/Group').default;
const SharePoint = require('../Icon/icons/SharePointLogoInverse').default;

const items = [
  <ActionLink text="Download" icon={Download} ariaLabel="Download this file" onClick={action('ActionLink clicked')} />,
  <ActionLink text="Upload a new version" icon={Upload} onClick={action('ActionLink clicked')} />,
  <ActionLink text="Group Info" icon={Group} ariaLabel="Go to group info page" href="https://www.yammer.com" newWindow={true} />,
  <ActionLink text="SharePoint Document Library website" icon={SharePoint} href="https://www.yammer.com" newWindow={true} />,
];

<div style={{ maxWidth: '200px' }}>
  <ActionLinkList items={items} />
</div>
```
