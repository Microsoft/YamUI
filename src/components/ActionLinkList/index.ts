/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default } from './ActionLinkList';
export * from './ActionLinkList.types';
