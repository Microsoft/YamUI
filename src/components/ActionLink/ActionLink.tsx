/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { join } from '../../util/classNames';
import Block, { TextSize } from '../Block';
import Clickable from '../Clickable';
import { FixedGridRow, FixedGridColumn, GutterSize } from '../FixedGrid';
import { BaseIcon, IconSize } from '../Icon';
import NavigationLink from '../NavigationLink';
import './ActionLink.css';

export interface BaseActionLinkProps extends BaseComponentProps {
  /**
   * The icon component class to be rendered.
   */
  icon: typeof BaseIcon;

  /**
   * A well-curated string describing the action for screen readers.
   */
  ariaLabel: string;

  /**
   * The visible text.
   */
  text: string;
}

export interface NavigationActionLinkProps extends BaseActionLinkProps {
  /**
   * A valid URL.
   */
  href: string;

  /**
   * Whether the link should open in a new window. It securely removes access to the opening window.
   */
  newWindow?: boolean;
}

export interface ClickableActionLinkProps extends BaseActionLinkProps {
  /**
   * A click handler.
   */
  onClick: ((e: React.MouseEvent<HTMLButtonElement | HTMLLinkElement>) => void);
}

export type ActionLinkProps = NavigationActionLinkProps | ClickableActionLinkProps;

/**
 * An actionable element with icon and text which will either render a button with an onClick callback
 * or a navigation link.
 */
export default class ActionLink extends React.Component<ActionLinkProps> {
  public render() {
    const { ariaLabel, className, icon: Icon, text } = this.props;
    const classNames = join(['y-actionLink', className]);

    const content = (
      <FixedGridRow gutterSize={GutterSize.SMALL} className="y-actionLink--wrapper">
        <FixedGridColumn fixed={true}>
          <Icon size={IconSize.MEDIUM} block={true} className="y-actionLink--icon" />
        </FixedGridColumn>
        <FixedGridColumn>
          <Block textSize={TextSize.MEDIUM_SUB}>{text}</Block>
        </FixedGridColumn>
      </FixedGridRow>
    );

    if ((this.props as NavigationActionLinkProps).href) {
      return (
        <NavigationLink
          href={(this.props as NavigationActionLinkProps).href}
          newWindow={(this.props as NavigationActionLinkProps).newWindow}
          ariaLabel={ariaLabel}
          className={classNames}
        >
          {content}
        </NavigationLink>
      );
    }
    return (
      <Clickable
        onClick={(this.props as ClickableActionLinkProps).onClick}
        ariaLabel={ariaLabel}
        className={classNames}
        block={true}
      >
        {content}
      </Clickable>
    );
  }
}
