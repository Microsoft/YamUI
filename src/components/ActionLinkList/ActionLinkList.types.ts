/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseComponentProps } from '../../util/BaseComponent/props';
import ActionLink from '../ActionLink';

export interface ActionLinkListProps extends BaseComponentProps {
  /**
   * An array of ActionLink components.
   */
  items: React.ReactElement<ActionLink>[];
}
