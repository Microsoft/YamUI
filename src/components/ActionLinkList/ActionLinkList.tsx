/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { LayoutList, LayoutListItem } from '../LayoutList';
import { ActionLinkListProps } from './ActionLinkList.types';
import ActionLink from '../ActionLink';
import { getClassNames } from './ActionLinkList.styles';
import { join } from '../../util/classNames';

/**
 * A vertical list of ActionLinks, with padding optimized for the sidebar or rightrail.
 */
export default class ActionLinkList extends React.Component<ActionLinkListProps> {
  public render() {
    const classes = getClassNames();

    return (
      <LayoutList direction="vertical" className={join(['y-actionLinkList', classes.root])}>
        {this.props.items.map(this.renderItem)}
      </LayoutList>
    );
  }

  private renderItem = (item: React.ReactElement<ActionLink>, index: number) => {
    return <LayoutListItem key={index}>{item}</LayoutListItem>;
  };
}
