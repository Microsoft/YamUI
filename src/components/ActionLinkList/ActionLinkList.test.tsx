/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import ActionLink from '../ActionLink';
import ActionLinkList, { ActionLinkListProps } from './index';
import Down from '../Icon/icons/Down';
import Up from '../Icon/icons/Up';

describe('<ActionLinkList />', () => {
  let component: ShallowWrapper<ActionLinkListProps>;

  describe('with items', () => {
    beforeEach(() => {
      const items = [
        <ActionLink text="Download" icon={Down} href="https://yammer.com" key="1" />,
        <ActionLink text="Upload" icon={Up} href="https://yammer.com" key="2" />,
      ];
      component = shallow(<ActionLinkList items={items} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
