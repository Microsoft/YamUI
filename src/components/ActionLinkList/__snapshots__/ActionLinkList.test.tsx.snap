// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionLinkList /> with items matches its snapshot 1`] = `
<LayoutList
  className=
      y-actionLinkList
      & .y-actionLink {
        padding-bottom: 0.6rem;
        padding-left: 0;
        padding-right: 0;
        padding-top: 0.2rem;
      }
  direction="vertical"
>
  <LayoutListItem
    key="0"
  >
    <ActionLink
      href="https://yammer.com"
      icon={[Function]}
      key="1"
      text="Download"
    />
  </LayoutListItem>
  <LayoutListItem
    key="1"
  >
    <ActionLink
      href="https://yammer.com"
      icon={[Function]}
      key="2"
      text="Upload"
    />
  </LayoutListItem>
</LayoutList>
`;
